src/gc/transaction_level_gc_manager.cpp

```diff
@@ -0,0 +1,307 @@
+      // it means that no active transactions can read it.
+      // so we can unlink it.
+      // we need to delete all the tuples from the indexes to which it belongs as well.
+      DeleteFromIndexes(garbage_ctx);
+      // Add to the garbage map
+      garbages.push_back(garbage_ctx);
+      tuple_counter++;
+
+    } else {
+      // if a tuple cannot be reclaimed, then add it back to the list.
+      local_unlink_queues_[thread_id].push_back(garbage_ctx);
+    }
+  }  // end for
+
+  auto safe_max_cid = concurrency::TransactionManagerFactory::GetInstance().GetNextCommitId();
+  for(auto& item : garbages){
+      reclaim_maps_[thread_id].insert(std::make_pair(safe_max_cid, item));
+  }
+  LOG_TRACE("Marked %d tuples as garbage", tuple_counter);
+}
+
+// executed by a single thread. so no synchronization is required.
+void TransactionLevelGCManager::Reclaim(const int &thread_id, const cid_t &max_cid) {
+  int gc_counter = 0;
+
+  // we delete garbage in the free list
+  auto garbage_ctx_entry = reclaim_maps_[thread_id].begin();
+  while (garbage_ctx_entry != reclaim_maps_[thread_id].end()) {
+    const cid_t garbage_ts = garbage_ctx_entry->first;
+    auto garbage_ctx = garbage_ctx_entry->second;
+
+    // if the timestamp of the garbage is older than the current max_cid,
+    // recycle it
+    if (garbage_ts < max_cid) {
+      AddToRecycleMap(garbage_ctx);
+
+      // Remove from the original map
+      garbage_ctx_entry = reclaim_maps_[thread_id].erase(garbage_ctx_entry);
+      gc_counter++;
+    } else {
+      // Early break since we use an ordered map
+      break;
+    }
+  }
+  LOG_TRACE("Marked %d txn contexts as recycled", gc_counter);
+}
+
+// Multiple GC thread share the same recycle map
+void TransactionLevelGCManager::AddToRecycleMap(std::shared_ptr<GarbageContext> garbage_ctx) {
+  
+  for (auto &entry : *(garbage_ctx->gc_set_.get())) {
+
+    auto &manager = catalog::Manager::GetInstance();
+    auto tile_group = manager.GetTileGroup(entry.first);
+
+    PL_ASSERT(tile_group != nullptr);
+
+    storage::DataTable *table =
+      dynamic_cast<storage::DataTable *>(tile_group->GetAbstractTable());
+    PL_ASSERT(table != nullptr);
+
+    oid_t table_id = table->GetOid();
+
+    for (auto &element : entry.second) {
+
+      // as this transaction has been committed, we should reclaim older versions.
+      ItemPointer location(entry.first, element.first); 
+      
+      // If the tuple being reset no longer exists, just skip it
+      if (ResetTuple(location) == false) {
+        continue;
+      }
+      // if the entry for table_id exists.
+      PL_ASSERT(recycle_queue_map_.find(table_id) != recycle_queue_map_.end());
+      recycle_queue_map_[table_id]->Enqueue(location);
+
+    }
+  }
+
+}
+
+// this function returns a free tuple slot, if one exists
+// called by data_table.
+ItemPointer TransactionLevelGCManager::ReturnFreeSlot(const oid_t &table_id) {
```

Comment:
This function is unused, which means we actually do not reuse any recycled tuple. Please revive the reuse logic in data_table.cpp.